### ISSUE TITLE
Common tags for ginkgo based tests, including functional(e2e) and unit tests

### ISSUE THREAD

**Issue by jeffvance** [kubevirt/containerized-data-importer#359]
This issue is here to start the definition of tags for all tests (functional and unit) that are based on the ginkgo framework. Ginkgo allows test filtering based on using regular expressions in the `skip=` and `focus=` test invocation args.  Eg. `ginkgo test ... -focus="\[e2e\]"`.

These tags can be simple/flat, eg _[e2e]_, _[importer]_, _[cloner]_, etc.  Or they can be hierarchical, eg: _[e2e/importer]_, _[e2e.clonerer]_, etc. However, there is no real need for hierarchical tags since multiple tags can be used in the ginko skip= and focus= args.

Proposal:
- [CLONER] - all cloner tests match this tag
- [DATAVOLUME] - all dataVolume tests match this tag.
- [E2E] - all functional tests match this tag
- [FLAKY] - tests that fail periodically usually for unknown reasons and require further debugging. Useful to skip these tests in CI. 
- [HW-DEP] - eg fast clone test would set this tag
- [IMPORTER] - all importer tests match this tag
- [SERIAL] - tests cannot be run in parallel
- [SLOW] - set of tests that have longer timeouts, sleep, or other factor that can result in slow completion times
- [SW-DEP] - tests that require specific libs installed on the node(s)
- [UNIT] - all unit tests match this tag
- [UPLOADER] - all uploader tests match this tag

An example of running only e2e cloner tests:
` ginkgo test ... -focus="\[E2E\]" -focus="\[CLONER\]"`

**Comment by jeffvance** [comment]
@nellyc @shiywang @copejon @screeley44 @davidvossel @awels

**Comment by nellyc** [comment]
Sounds good! 
Just 2 comments:
* I think that 'E2E' should be renamed to 'Functional'
* I dont really like the idea of a Flaky tests tag, as we may end up forgetting about them ;) in my POV a flaky test should be either fixed ASAP or redesigned if we cannot fix it (basically we dont get any value from such tests a)

@fabiand what do you say about similar tags for kubevirt? (excluding feature specific ones)

**Comment by awels** [comment]
One more to add [OC] means its an openshift specific test, that requires either an open shift client or oc instead of kubectl.

To help polarion we would also like to put on context or it level another 'tag' that is more complex that contain a set of values to help polarion organize the tests. The format would look something like this:
[rfe_id: xxx, crit: (low/med/high), level: component, component: importer, type: (positive/negative)]

The polarion report can parse this label and populate fields that are important to it from there.

**Comment by aglitke** [comment]
Seems okay to me.  @jeffvance Could you create a PR that documents the test tags we agree to use somewhere in the repo?

**Comment by copejon** [comment]
@jeffvance Do `[SW-DEP]` and `[HW-DEP]` have any overlap with tags other than `[CLONER]`? 

Some have made the point that we don't want parallel tests (citing encapsulation and debugging concerns) so `[SERIAL]` may be globally applicable for a long time.

@nellyc I've been curious about how kubevirt defines e2e.  Running tests that execute start to finish work flows in a cluster seems more end to end than functional.

**Comment by copejon** [comment]
@awels 
> To help polarion we would also like to put on context or it level another 'tag' that is more complex that contain a set of values to help polarion organize the tests. The format would look something like this:
[rfe_id: xxx, crit: (low/med/high), level: component, component: importer, type: (positive/negative)]
>
> The polarion reporter can parse this label and populate fields that are important to it from there

That could become pretty unwieldily as code and code coverage scales.  Is there another way we can enable polarion integration here?

**Comment by awels** [comment]
We have a [gingko report](https://github.com/kubevirt/qe-tools) but it is missing a few key fields that polarion requires. So we are brain storming ways of getting those fields into the reporter struct from the tests. The reporter has access to the strings in the Describe/It/Context so I was suggesting that as one way of passing the fields to the reporter.

I haven't looked too closely at what a gingko reporter has access to, maybe something else is available.

**Comment by jeffvance** [comment]
I  think I understand @nellyc 's concerns that if a `Describe` string changes, and she keys off of the original string then polarian will have essentially duplicate tests: 1) the original and 2) a duplicate test where the only difference is the name of the test. Is there a simpler

**Comment by jeffvance** [comment]
@awels ok, I don't know about polarian's use of these fields. I guess a point I was trying to make above is, imo, we should not sway cdi testing so far toward polarian that it causes test tagging to be unintuitive and/or high maintenance.

**Comment by fabiand** [comment]
Also speaking for the kubevirt side - I do understand that there is need for downstream distributions to classify and categorize tests.

I'd be fine with this -as I hope that kubevirt and friends will benefit from interest of downstreams.
My request would be that any metadata would contain a vendor or ownership information, ii.e. alongside what @nellyc wrote:

```
[rfe_id: xxx, crit: (low/med/high), type: negative, vendor:cnv-qe.redhat.com]
```

We'd also need to decide how this can be implemented (just comments, or annotations).

**Comment by fabiand** [comment]
> [rfe_id: xxx, crit: (low/med/high), type: negative, vendor:cnv-qe.redhat.com]

Or a better identifier which also allows a contact:
```
[rfe_id: xxx, crit: (low/med/high), type: negative, owner:cnv-qe@redhat.com]
```

**Comment by aglitke** [comment]
@nellyc Can this issue be closed now?

**Comment by nellyc** [comment]
@aglitke yes, finally we have the tests attributes on most tests \o/ 
we will continue to work to mark unmarked tests as well as new tests.
I am going to open a new ticket to support auto creation of test ids, so we will be able to automate the entire process

**Comment by shweta-bijay** [comment]
for our case, we have a git repo that contains only E2E tests, however, we want to tag tests as below so as to easily identify which test case failed. This can also help run specific category of tests. Does it make sense to support it OR is there some plugin that we can consider:
1. P0 / P1 / P2 / P3
2. Add team alias
3. Add feature name
